### PR TITLE
Support returning an integer from a Simple Program

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to #41704.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/SimpleProgramsTests.cs [hunk 1]
```diff
@@ -6233,5 +6237,333 @@ public void MissingTypes_02()
...
+                );
+        }
+
+        [Fact]
+        public void Return_01()
+        {
+            var text = @"
+System.Console.WriteLine(""Hi!"");
+return;
+";
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SimpleProgramsTests.cs:6364]: Are we testing unreachable code, after `return;` for instance?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/SimpleProgramsTests.cs:6364]: > Are we testing unreachable code, after return; for instance?

Not specifically. There is a test that verifies that we run flow analysis for top level statements. Or do you have something else in mind?

---
In reply to: [410414111](https://github.com/dotnet/roslyn/pull/43451#discussion_r410414111) [](ancestors = 410414111)

File: src/Compilers/CSharp/Test/Semantic/Semantics/SimpleProgramsTests.cs [hunk 2]
```diff
@@ -6233,5 +6237,333 @@ public void MissingTypes_02()
...
+            var entryPoint = SimpleProgramNamedTypeSymbol.GetSimpleProgramEntryPoint(comp);
+            Assert.Equal("System.Int32", entryPoint.ReturnType.ToTestDisplayString());
+            Assert.False(entryPoint.ReturnsVoid);
+            comp.VerifyDiagnostics(
+                // (4,5): error CS0126: An object of a type convertible to 'int' is required
+                //     return;
+                Diagnostic(ErrorCode.ERR_RetObjectRequired, "return").WithArguments("int").WithLocation(4, 5)
+                );
+        }
+
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SimpleProgramsTests.cs:6567]: Perhaps test `return default;` and `await Task.Yield(); return default;`.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs [hunk 3]
```diff
@@ -33,20 +33,29 @@ internal SynthesizedSimpleProgramEntryPointSymbol(SimpleProgramNamedTypeSymbol c
...
-            {
-                _returnType = Binder.GetSpecialType(containingType.DeclaringCompilation, SpecialType.System_Void, NoLocation.Singleton, diagnostics);
+                case (true, false):
+                    _returnType = Binder.GetWellKnownType(containingType.DeclaringCompilation, WellKnownType.System_Threading_Tasks_Task, diagnostics, NoLocation.Singleton);
+                    break;
+                case (false, false):
+                    _returnType = Binder.GetSpecialType(containingType.DeclaringCompilation, SpecialType.System_Void, NoLocation.Singleton, diagnostics);
+                    break;
+                case (true, true):
+                    _returnType = Binder.GetWellKnownType(containingType.DeclaringCompilation, WellKnownType.System_Threading_Tasks_Task_T, diagnostics, NoLocation.Singleton).
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs:47]: >NoLocation.Singleton [](start = 153, length = 20)

Should this be the location of the first `return` expression?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs [hunk 4]
```diff
@@ -33,20 +33,29 @@ internal SynthesizedSimpleProgramEntryPointSymbol(SimpleProgramNamedTypeSymbol c
             _declaration = declaration;
 
             bool hasAwait = declaration.HasAwaitExpressions;
+            bool hasReturnWithExpression = declaration.HasReturnWithExpression;
 
-            if (hasAwait)
+            switch (hasAwait, hasReturnWithExpression)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs:38]: >switch [](start = 12, length = 6)

Consider writing as a switch expression.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs:38]: > Consider writing as a switch expression.

I like switch statement better.

---
In reply to: [410423869](https://github.com/dotnet/roslyn/pull/43451#discussion_r410423869) [](ancestors = 410423869)

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs [hunk 5]
```diff
@@ -539,8 +539,14 @@ internal static bool HasYieldOperations(SyntaxNode? node)
         {
             // Do not descend into functions and expressions
             return node is object &&
-                   node.DescendantNodesAndSelf(child => !IsNestedFunction(child) && !(node is ExpressionSyntax)).
-                       OfType<YieldStatementSyntax>().Any();
+                   node.DescendantNodesAndSelf(child => !IsNestedFunction(child) && !(node is ExpressionSyntax)).Any(n => n is YieldStatementSyntax);
+        }
+
+        internal static bool HasReturnWithExpression(SyntaxNode? node)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs:545]: >SyntaxNode? [](start = 53, length = 11)

When is this `null`?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs:47]: > Should this be the location of the first return expression?

I do not believe this edge case is worth the effort locating the one return statement with expression.

---
In reply to: [410423382](https://github.com/dotnet/roslyn/pull/43451#discussion_r410423382) [](ancestors = 410423382)

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs:545]: > When is this null?

When the method is called with one.

---
In reply to: [410424982](https://github.com/dotnet/roslyn/pull/43451#discussion_r410424982) [](ancestors = 410424982)

File: src/Compilers/CSharp/Test/Semantic/Semantics/SimpleProgramsTests.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/SimpleProgramsTests.cs:6364]: Added tests

---
In reply to: [410417194](https://github.com/dotnet/roslyn/pull/43451#discussion_r410417194) [](ancestors = 410417194,410414111)

**Comment by AlekseyTs** [comment]
@cston I think I responded to your comments

**Comment by AlekseyTs** [comment]
@cston, @RikkiGibson, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@RikkiGibson, @dotnet/roslyn-compiler Please review, need a second sign-off.

**Pull request merged**